payu/models/staged_cable.py

```diff
@@ -238,52 +190,85 @@ def _apply_stage_namelists(self, stage_name):
-            # If the current stage is a non-empty string, it means we exited
-            # during the running of the previous stage- leave as is
-            stage_name = self.configuration_log['current_stage']
-        else:
-            # Normal case where we just archived a successful stage.
-            self.configuration_log['current_stage'] = \
-                    self.configuration_log['queued_stages'].pop(0)
+        self.configuration_log['current_stage'] = \
+            self.configuration_log['queued_stages'].pop(0)
 
         self._save_configuration_log()
-
-        # Copy the log to the work directory
-        shutil.copy('configuration_log.yaml', self.work_input_path)
+        conf_log_p = os.path.join(self.control_path, 'configuration_log.yaml')
+        shutil.copy(conf_log_p, self.work_path)
 
     def archive(self):
         """Store model output to laboratory archive and update the
 configuration log."""
 
-        # Move files from the restart directory within work to the archive
-        # restart directory.
+        # Retrieve all the restarts required for the next stage
+        self._collect_restarts()
+
+        # Update the configuration log and save it to the working directory
+        self._read_configuration_log()
+        self._archive_current_stage()
+
+        # Now set the number of runs using the configuration_log
+        remaining_stages = len(self.configuration_log['queued_stages'])
+        print("Overriding the remaining number of runs according to the " +
+              "number of queued stages in the configuration log.")
+        self.expt.n_runs = remaining_stages
+
+        conf_log_p = os.path.join(self.control_path, 'configuration_log.yaml')
+        if self.expt.n_runs == 0:
+            # Configuration successfully completed
+            os.remove(conf_log_p)
+
+        super(StagedCable, self).archive()
+
+    def _collect_restarts(self):
+        """Collect all restart files required for the next stage. This is a
+        merge of the files in work_path/restart and in prior_restart_path, with
+        the files in work_path/restart taking precedence."""
+
+        # First, collect restarts which do not have a newer version (when the
+        # counter is greater than 0)
+        if self.expt.counter > 0:
+            prior_restart_dir = 'restart{0:03}'.format(self.expt.counter - 1)
+            prior_restart_path = os.path.join(self.expt.archive_path,
+                                              prior_restart_dir)
+
+            # For each restart, check if newer version was created. If not,
+            # copy into the work restart path.
+            generated_restarts = os.listdir(self.work_restart_path)
+
+            for f in os.listdir(prior_restart_path):
+                if f not in generated_restarts:
+                    shutil.copy(os.path.join(prior_restart_path, f),
+                                self.work_restart_path)
+
```

Comment:
I think it should be possible to just make links, which will resolve to their original location IIRC.

Reply:
What about the scenario where we want to share a restart directory with someone else, or perhaps even publicly somewhere, so that they can use it as a start point for simulations? I think it's likely enough that the original run would be somewhere that others won't have read permissions to, or even potentially on other machines.

Reply:
I haven't looked into the code changes but I don't see how making a copy solves problems of read permissions? You need read access to be able to copy files. So links would just be as well.

Reply:
I'm thinking of a situation where someone (e.g. ACCESS-NRI) has run an expensive spin-up configuration and want to provide the restarts as a starting point for others in the community. The configuration may have 20+ stages, but the full suite of restarts may contain a restart from the 1st stage, one from the 5th stage and the rest from the most recent stage. A supported ```payu``` workflow is to take a restart from elsewhere and effectively plug it in, so ```payu``` transparently restarts from this restart (I think this is true- I have memory of this being demonstrated on the training day).

I'm imagining a scenario where I run a configuration to create a starting point for community science. I'll likely save the data to my non-permanent scratch space. I want to take the most recent ```payu``` set of restarts and store it someone accessible on ```gdata```. If the necessary older restarts are only known by symlink, then they'll disappear relatively soon and whichever user takes this ```payu``` restart may not have access to my scratch space where the symlinked files live.

Reply:
It's because the `archive` space is created under `/scratch` by default and it's the user's responsibility to save it to `/g/data` if required. I'd forgotten about that bit for payu.

So copy makes sense.

Reply:
You can use `cp` with the `-L` option to "dereference" links, so the destination will contain all the files referenced by links

https://superuser.com/a/216920